rfcs/text/0000-metadata-integration-test.md

```diff
@@ -0,0 +1,78 @@
+* [Sam Pierson](mailto:spierson@chanzuckerberg.com)
+* [Matt Weiden](mailto:mweiden@chanzuckerberg.com)
+* [Mark Diekhans](mailto:markd@ucsc.edu)
+* [Hannes Schmidt](mailto:hannes@ucsc.edu)
+
+## Shepherd
+
+## Motivation
+
+Currently breakages to DCP systems caused by metadata schema changes are detected at runtime upon data (bundle) upload. This forces data wranglers, DCP operators, and data consumers to react to unpredictable runtime failures rather than detecting problems before the code is merged and deployed.
+
+To alleviate this issue, the DCP development team can use a metadata schema integration test to test whether it is safe to release data downstream of the Data Storage Service (DSS), past which point applications start to depend on the schema. The benefits of this approach are that it will:
+
+1. Make the development environment more stable
+1. Make updates to software and schema smaller and more frequent
+1. Make failures due to metadata schema changes more readily observable
+1. Increase availability of DCP systems downstream of DSS
+
+## Detailed Design
+
+In this design we will use the following terms:
+ 
+ * "schema integration test system" - an automated CI/CD pipeline used to check that downstream systems can handle bundles with a new schema
+ * "new schema label" - anything that indicates that the schema has changes that have not yet been tested against downstream systems: potentially a JSON field, git tag, git branch, or comparison to tested version list
+
+
+Design
+
+1. Bundles with new metadata schema changes are uploaded to the production environment with a new schema label
+1. Observing the new schema label, the DSS stores the new bundles and makes them available with the `GET /v1/bundles` endpoint with the bundle `uuid` or `(uuid, version)`, but does not index nor does it release subscription notifications for them.
+1. Via a schema integration test system, a sample of the uploaded bundles are copied from the production DSS to the integration DSS testing environment.
+1. The integration DSS deployment stores the bundles, indexes their metadata, and releases bundle events to downstream systems in the integration environment with subscriptions.
+1. The schema integration test checks the results and passes if systems downstream of the integration DSS have correctly processed the new bundles; it fails otherwise.*
+1. If the schema integration test passes, the new bundles in the production DSS can be released downstream by issuing a new bundle version of each without the new schema label. This process could be automated.
+1. If the schema test does not pass, the data is not released and development teams are notified to resolve the issue and rerun the test.
+
+![Integration test design](../images/0000-integration-test.png)
+
+*Integration test data flow: test path is shown in red, production release in blue, integration tooling in dashed lines; edges are labeled with corresponding step numbers in the design above*
```

Comment:
I think we need to avoid coupling the consumption by pipelines and consumption by browser.  There will be cases where new types of pipelines will need to run that require newer versions of a particular schema that the browser will not be able to handle. 

If we stick with this design, we should make it clear in the cons that analysed data can't be released to downstream consumers until it passes the browser tests.

Reply:
I agree with @simonjupp on this. From my perspective, I see four "data flow" paths.
1. Ingest -> DS (submit primary data)
1. DS -> Browser (display primary data)
1. DS -> Pipelines -> Ingest -> DS (analyze primary data and submit secondary data)
1. DS -> Browser (display secondary data)

(1) should happen even if (2) - (4) fail. (2) should happen even if (3) and (4) fail (or aren't supposed to happen). (3) should happen even if (4) fails. There might be other "ifs" I'm missing, but the general idea is that failing downstream paths shouldn't prevent upstream paths from happening. I think some amount of retries/replays are necessary as we are always going to be playing "catch up" with new data types being submitted.

Reply:
To echo @malloryfreeberg's point - I'd question whether we really want to prevent the release of raw or analysis data to tertiary portals ("red boxes") because e.g. data cannot be correctly displayed in the browser. I appreciate the technical reasons as to why this might be difficult, so it's sensible to be pragmatic here, but the limitations of this design at least need to be called out.

Reply:
@tburdett @diekhans @simonjupp Per our discussion at the F2F, I've added the following sentence/design element. I believe it addresses this concern.

> The DSS subscription API should enable clients to opt into receiving events with unsupported metadata schema versions. Notably, this would enable the data browser to display data with unsupported metadata schema versions even if analysis cannot yet process it.

Note, this doesn't change the rest of the steps. If the data browser can't display the data, there's no sense in releasing a new version of it.